typeclass-experiments/src/Gen/Rust.hs

```diff
@@ -0,0 +1,161 @@
+  List t -> do
+    t' <- runGenerator scope t
+    pure $ "Vec<" <> t' <> ">"
+  other -> matchFail (List _x) other
+
+{- Records -}
+rustRecFields :: Gen
+rustRecFields = Gen (RecFields _l _x _xs) $ \scope pat -> case pat of
+  RecFields (Name n) t Nil -> do
+    t' <- runGenerator scope t
+    pure $ indent2 $ n <> ": " <> t' -- <> ","
+  RecFields (Name n) t ys -> do
+    t' <- runGenerator scope t
+    rest <- runGenerator scope (RecP ys)
+    pure $ indent2 $ n <> ": " <> t' <> "," <> "\n" <> rest
+  other -> matchFail (RecFields _l _x _xs) other
+
+rustRecConstr :: Gen
+rustRecConstr = Gen (RecBody _l _xs) $ \scope pat -> case pat of
+  RecBody (Name nm) fields -> do
+    body <- runGenerator scope (RecP fields)
+    pure $ nm <> " " <> brackets body
+  other -> matchFail (RecBody _l _xs) other
+
+rustBareType :: Gen
+rustBareType = Gen (ProdP (_x :* Nil)) $ \scope pat -> case pat of
+  ProdP (t :* Nil) -> runGenerator scope t
+  other -> matchFail (ProdP (_x :* Nil)) other
+
+rustTupleBody :: Gen
+rustTupleBody = Gen (ProdP (_a :* _x :* _xs)) $ \scope pat -> case pat of
+  ProdP (t :* ts) -> do
+    t' <- runGenerator scope t
+    ts' <- runGenerator scope (ProdP ts)
+    pure $ t' <> ", " <> ts'
+  other -> matchFail (ProdP (_x :* _xs)) other
+
+rustTupleConstr :: Gen
+rustTupleConstr = Gen (ProdBody _l _xs) $ \scope pat -> case pat of
+  ProdBody (Name cName) ys -> do
+    ys' <- runGenerator scope (ProdP ys)
+    pure $ cName <> parens ys'
+  other -> matchFail (_l := ProdP _xs) other
+
+rustTypeNoFreeVars :: Gen
+rustTypeNoFreeVars = Gen (Sum _name _nil _x) $ \scope pat -> case pat of
+  Sum (Name tName) Nil p -> case p of
+    -- normal record struct, will be caught by rustRecConstr
+    t@(SumP (_ := RecP _ :* Nil)) -> do
+      struct <- runGenerator scope t
+      pure $ "struct " <> struct <> ";\n"
+
+    -- tuple struct, will be caught by rustTupleConstr
+    t@(SumP (_ := ProdP _ :* Nil)) -> do
+      tup <- runGenerator scope t
+      pure $ "struct " <> tup <> ";\n"
+    SumP ts@(_ :* _) -> do
+      case pToList ts of
+        Nothing -> Left $ MalformedPatList ts
+        Just tsList -> do
+          constructors <- mapM (runGenerator scope) tsList
+          let enumBody = brackets (T.concat $ map (\x -> indent x <> ",\n") constructors) <> "\n"
+          pure $ "\nenum " <> tName <> " " <> enumBody
+    other -> matchFail (Sum _name _nil _x) other
+  other -> error $ show other
+  where
+    pToList :: Pat -> Maybe [Pat]
+    pToList = \case
+      -- \/ NOTE: Fix rustRecConstr and rustTupleConstr so this stupid thing isn't necessary
+      (z :* zs) -> (SumP (z :* Nil) :) <$> pToList zs
+      Nil -> Just []
+      _ -> Nothing
+
+rustScope :: Set Gen
+rustScope =
+  S.fromList
+    [ rustInt
+    , rustString
+    , rustBool
+    , rustRef
+    , rustMaybe
+    , rustList
+    , rustBareType
+    , rustRecFields
+    , rustRecConstr
+    , rustTupleBody
+    , rustTupleConstr
+    , rustTypeNoFreeVars
+    ]
```

Comment:
This is quite confusing, I was going through this thinking it's all about typeclasses, and this is all about type definition mapping? Is that somehow covered or did I just misread?

Reply:
I guess I could have been clearer in the initial comment to this PR, so lemme try again. This library contains 3 related things: 

 - A representation of patterns for our `Source` type. (`Common.Types`)
 - A constraint solver that makes use of those patterns to solve typeclass constraints. (`Resolve._`)
 - A code generation framework that uses those patterns to allow for modular definitions of code generation functions. (`Gen._`)
 
Because we need patterns anyway to do anything useful with typeclass constraints, I thought it would be useful to explore their relevance to generating code for the types themselves. Modulo some awkwardness resulting from using `Text` instead of `Doc` I think it actually works quite nicely? 

The idea is: If we're going to match on patterns of our SOP-like representation to generate/derive typeclass instances, it makes a lot of sense to use the same machinery to generate the type definitions in a target language as well.

Reply:
That sounds great